contentprovider.go

```diff
@@ -261,41 +261,49 @@ func (nls newlines) atOffset(offset uint32) (lineNumber, lineStart, lineEnd int)
 		return nls.locs[n] >= offset
 	})
 
-	end := int(nls.fileSize)
-	if idx < len(nls.locs) {
-		end = int(nls.locs[idx])
+	start, end := nls.lineBounds(idx + 1)
+	return idx + 1, int(start), int(end)
+}
+
+// lineBounds returns the byte offsets of the start and end of the 1-based
+// lineNumber. The end offset is exclusive and will not contain the line-ending
+// newline. If the line number is out of range of the lines in the file, start
+// and end will be clamped to [0,fileSize].
+func (nls newlines) lineBounds(lineNumber int) (start, end uint32) {
+	// newlines[0] is the start of the 2nd line in data.
+	startIdx := lineNumber - 2
+	endIdx := lineNumber - 1
+
+	if startIdx >= 0 && startIdx < len(nls.locs) {
+		start = nls.locs[startIdx] + 1
+	} else if startIdx < 0 {
+		start = 0
+	} else if startIdx >= len(nls.locs) {
+		start = nls.fileSize
 	}
 
-	start := 0
-	if idx > 0 {
-		start = int(nls.locs[idx-1] + 1)
+	if endIdx >= 0 && endIdx < len(nls.locs) {
+		end = nls.locs[endIdx]
+	} else if endIdx < 0 {
+		end = 0
+	} else if endIdx >= len(nls.locs) {
+		end = nls.fileSize
 	}
```

Comment:
```
start = nls.lineStart(lineNumber - 2)
end = nls.lineStart(lineNumber - 1)
```

with the corresponding method, now that we have the newlines type

?

Reply:
Would also simplify the call sites
(currently
```
lowStart, _ := nls.lineBounds(low)
	_, highEnd := nls.lineBounds(high - 1)
```
)

 and probably remove the need of this method

Reply:
The problem with this is that, in general, `lineEnd(n) != lineStart(n+1)`. This is because we do not include the newline character in the line bounds. 

We might be able to safely say `lineStart := lineEnd(n) + 1`, but only because we clamp offsets returned by `lineStart` to `[0, len(fileContent))`. However, I think it's preferable to keep this behavior a more explicit part of the method call API than to depend on callers to remember to do this. Especially since it takes some convincing to be confident that that is always right in all edge cases (file boundaries, empty lines, no trailing newline, etc.)

Reply:
Now I see that they differ by 1. Ok.